featurebyte/routes/common/feature_metadata_extractor.py

```diff
@@ -72,18 +80,43 @@ async def extract_from_object(self, obj: BaseFeatureModel) -> dict[str, Any]:
 
         Returns
         -------
-        dict[str, Any]
+        tuple[GroupOperationStructure, dict[str, Any]]
         """
         try:
             op_struct = obj.extract_operation_structure(keep_all_source_columns=False)
-            return await self._extract(op_struct=op_struct)
+            metadata = await self._extract(op_struct=op_struct)
         except KeyError:
             # FIXME (https://featurebyte.atlassian.net/browse/DEV-2045): Setting
             #  keep_all_source_columns to False fails in some edge cases. This is a workaround for
             #  the cases where it fails, though it produces info that is more verbose than desired.
             #  We should clean this up.
             op_struct = obj.extract_operation_structure(keep_all_source_columns=True)
-            return await self._extract(op_struct=op_struct)
+            metadata = await self._extract(op_struct=op_struct)
+        return op_struct, metadata
+
+    @classmethod
+    def _reference_key_func(
+        cls,
+        col: Optional[
+            Union[SourceDataColumn, DerivedDataColumn, AggregationColumn, PostAggregationColumn]
+        ],
+    ) -> str:
+        # helper function to generate reference key used in the reference map (_extract function)
+        if col is None:
+            return "None"
+        if isinstance(col, SourceDataColumn):
+            return f"SourceDataColumn({col.table_id}, {col.name})"
+        if isinstance(col, DerivedDataColumn):
+            table_ids = sorted([col.table_id for col in col.columns if col.table_id is not None])
+            return f"DerivedDataColumn({col.name}, {table_ids})"
+        if isinstance(col, AggregationColumn):
+            # for count aggregation, col.column is None
+            col_key = cls._reference_key_func(col.column) if col.column is not None else None
+            return f"AggregationColumn({col.name}, {col_key})"
```

Comment:
As the `col` contains other attribute like `filter` to track whether the column has been applied filtering operation, it could cause referencing error when both `col` & filtered `col` are used in deriving a column. A referencing key function is introduced to use the source table(s) & column name information only to construct the reference key.

Reply:
Will update this in the coming PR.